### PR TITLE
Add trustee-agent - a simple tool to fetch secrets from Trustee

### PR DESCRIPTION
**Pull request by uril**
The trustee attester is a tool to attest and fetch secrets from Trustee.

Trustee attester is using attestation-agent's kbs_protocol client and attesters to gather hardware-based confidential-computing evidence and send it over to Trustee.

This PR replaces #755. Binary renamed and moved under kbs_protocol.
Also Trustee URL and optional certificate are given in the command line and not in a configuration file.

### CODE REVIEW

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:15]:
```suggestion
This can be formatted into a markdown style

Build with:
```bash
cargo build -p kbs_protocol --bin trustee-agent --no-default-features \
          --features "background_check,passport,<openssl|rusttls>,bin,<attesters-list>"
```
```

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:19]: Also this

```suggestion
```bash
./target/debug/trustee-agent --url <Trustee-URL> [--cert-file <path>] get-resource --path <resource-path>
```
```

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:22]:
```suggestion
```bash
./target/debug/trustee-agent --url http://localhost:50000 get-resource --path default/keys/dummy
```
```

**Review by Xynnn007 — COMMENTED** [review]
Thanks @uril ! The code looks pretty good and I like it! Some last comments from my side

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:19]: There should also be another ```   the next line. It can not be correctly printed here. THe same to the following

**Comment by Xynnn007** [comment]
`cargo fmt` needs to be performed under the codebase to make the CI happy.

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (mkulke) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs:80]: I don't think we need this json conversion. can you create a uri with `kbs_protocol::ResourceURI::new()?`

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (mkulke) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs:83]: I'd make this configurable. base64 by default is fine, but sometimes you want to pipe the output into another process or a file and then it's clunky

**Review by mkulke — COMMENTED** [review]
some nits

is "agent" the appropriate name here? i'd associate agent with a long-running process that is doing something on behalf of me, like attestation-agent or ssh-agent

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (beraldoleal) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs:42]: This repo is archived, maybe use this instead?

https://github.com/confidential-containers/guest-components/blob/main/attestation-agent/docs/KBS_URI.md

File: attestation-agent/kbs_protocol/Cargo.toml
Comment (uril) [reply attestation-agent/kbs_protocol/Cargo.toml]: Yes, done.

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md]: Thanks for the suggestion. I added this and some headers

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md]: I'm not sure what you mean with "another ``` the next line."
I added the ```bash to the patch.

I left it without "./target/debug/" as the program can be also run from e.g. /usr/bin or /usr/local/bin (after cargo install)

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md]: same as above

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs]: Fixed, thanks.

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs]: You're correct. I replaced it with new() as suggested.

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs:86]: This configuration can be added later.
If not printed as base64, it's printed as a vector of bytes.
I can print it as String, and it may be more convenient to the user, but my concern is
that some bytes may be non-printable.
// let secret = String::from_utf8(resource_bytes)?;

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs]: As for the name, the program does do the attestation and resource retrieval on your behalf, but it's not a long running process. It runs on the guest, similar to the attestation-agent (which is a long lived service).
I suggested the name trustee-client, but it is already taken for management of Trustee.
The name trustee-resource-getter was also suggested and is most accurate but it's long.
Would you prefer one of trustee-kbc, trustee-cc-kbc ?

**Comment by uril** [comment]
> `cargo fmt` needs to be performed under the codebase to make the CI happy.

Thanks @Xynnn007, it found two issues.

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:15]:
```suggestion
## Build:
```

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:1]:
```suggestion
# Trustee agent
```

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:22]:
```suggestion
$ trustee-agent --url <Trustee-URL> [--cert-file <path>] get-resource --path <resource-path>
```

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md:27]:
```suggestion
## Example
$ trustee-agent --url http://localhost:50000 get-resource --path default/keys/dummy
```

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs:80]: If the `resource_url` here is always an empty string, we can directly use `""` (`&str`) instead of always "new" a string and refer it
```suggestion
            let resource_path = format!("/{}", path); // path must start with a '/'
            let resource = ResourceUri::new("", &resource_path)?;
```

**Review by Xynnn007 — COMMENTED** [review]
Thanks. Some more nits about the readme document. The final rendering shape can be seen on the page https://github.com/uril/coco-guest-components/tree/trustee-agent/attestation-agent/kbs_protocol/src/bin/trustee-agent. You can take a look if it meets your ideal shape when debugging

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (mkulke) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs:82]: yeah, I think it's correct to output b64 by default, but it would be useful if we had a `--raw` flag or something, in case we want to pipe/redirect the secret somewhere. It's not required, though.

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (mkulke) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs]: I'm uninspired, too. "trustee-key-release" or "trustee-util" maybe?

Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs]: How about  "trustee-attester" ?

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md]: Yes, thanks, I fixed it with ## BUILD: ##

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md]: ````bash`   starts the block and ````` closes it

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md]: Fixed, thanks

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/README.md]: I added ```   "close" the code-block

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs]: I've renamed it, thanks

**Comment by uril** [comment]
> Thanks. Some more nits about the readme document. The final rendering shape can be seen on the page https://github.com/uril/coco-guest-components/tree/trustee-agent/attestation-agent/kbs_protocol/src/bin/trustee-agent. You can take a look if it meets your ideal shape when debugging

Thanks, that was a good tip. I fixed the markdown syntax. 
Note that the rename broke this link.

File: attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-agent/main.rs]: Okay

File: attestation-agent/kbs_protocol/src/bin/trustee-attester/README.md
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-attester/README.md:1]: The preview looks good, while we do not need the `#` in the end of the line when making headings. https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#headings

File: attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs
Comment (Xynnn007) [inline attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs:25]: Is this something left for debug?

**Review by Xynnn007 — APPROVED** [review]
Last comments, otherwise looks good to me. Please pay attention to @mkulke 's comments

File: attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs
Comment (mkulke) [inline attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs:79]: nit: this will bail with an obscure error message if the user-specified path already starts with a `/`, so I would suggest to either:

a) validate the path format with a regex, so we can safely prepend the /

b) prepend the / only if it's not present: `format!("/{}", path.trim_start_matches('/'))` (even if it looks a bit ridiculous)

**Review by mkulke — APPROVED** [review]
a small nitpick, other lgtm

File: attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs
Comment (uril) [reply attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs]: Yes, removed.

File: attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs
Comment (mkulke) [reply attestation-agent/kbs_protocol/src/bin/trustee-attester/main.rs]: 👍 for booleans we can stick to if/else statements though, they also are expressions in rust, i.e. you can return from a branch: let x = if 1 != 2 { "foo" } else { "bar" }

**Review by mkulke — CHANGES_REQUESTED** [review]
sorry I just noticed this: we shouldn't merge the PR without CI for the tool.

we don't need to publish binaries, but there should at least be compile/lint/format checks for the code.

you can take a look at the AA build for reference: https://github.com/confidential-containers/guest-components/blob/main/.github/workflows/aa_basic.yml

**Comment by uril** [comment]
Added a commit with CI. Thanks for recommending aa_basic.yml.

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [inline .github/workflows/trustee-attester.yml:27]: why this limitation?

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [inline .github/workflows/trustee-attester.yml:41]: i don't think this is a matrix. we only run the suite on ubuntu 24.04 w/ rust stable

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [inline .github/workflows/trustee-attester.yml:64]: we only run on ubuntu 24.04

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [inline .github/workflows/trustee-attester.yml:70]: we only run this on ubuntu 24.04

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [inline .github/workflows/trustee-attester.yml:41]: why --no-default-features?

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [inline .github/workflows/trustee-attester.yml:81]: if we want to share the opts, we can set an ENV on the job like this:

image

the env can be reference as `${{ env.CARGO_TEST_OPTS }}`

File: .github/workflows/trustee-attester.yml
Comment (uril) [reply .github/workflows/trustee-attester.yml:27]: Same as in aa_basic.yml

File: .github/workflows/trustee-attester.yml
Comment (uril) [reply .github/workflows/trustee-attester.yml:41]: I removed the s390x entry of the list, for now, so only ubuntu is left there.
Possibly someone will re-add it later.

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [reply .github/workflows/trustee-attester.yml:27]: I think that's a bug. let's remove it here.

File: .github/workflows/trustee-attester.yml
Comment (mkulke) [reply .github/workflows/trustee-attester.yml:41]: I see, maybe add this as a comment above the matrix statement. otherwise, it's unclear to the reader of that workflow.

Comment (uril) [reply .github/workflows/trustee-attester.yml:41]: Build only with the " --features" that are specified.
Same for dependent crates, only use features that are specified.

Comment (uril) [reply .github/workflows/trustee-attester.yml]: I re-added s390x. It was added to many other .yml files

**Review by mkulke — APPROVED** [review]

**Comment by fitzthum** [comment]
@uril please keep an eye on the new workflow after we merge this. We may need a follow-up with some fixes.

**Pull request merged**
